src/engine/LoadURL.cpp

```diff
@@ -0,0 +1,167 @@
+
+#include "global/RuntimeParameters.h"
+#include "util/http/HttpUtils.h"
+
+// _____________________________________________________________________________
+string LoadURL::getCacheKeyImpl() const {
+  // TODO<qup42> do caching based on ETag, Last-Modified or similar
+  if (RuntimeParameters().get<"cache-load-results">()) {
+    return absl::StrCat("LOAD URL ", loadURLClause_.url_.asString(),
+                        loadURLClause_.silent_ ? " SILENT" : "");
+  }
+  return absl::StrCat("LOAD URL ", cacheBreaker_);
+}
+
+// _____________________________________________________________________________
+string LoadURL::getDescriptor() const {
+  return absl::StrCat("LOAD URL ", loadURLClause_.url_.asString());
+}
+
+// _____________________________________________________________________________
+size_t LoadURL::getResultWidth() const { return 3; }
+
+// _____________________________________________________________________________
+size_t LoadURL::getCostEstimate() {
+  // This Operation will always be the only operation in a query tree, so we
+  // don't really need estimates.
+  return 10 * getSizeEstimateBeforeLimit();
+}
+
+// _____________________________________________________________________________
+uint64_t LoadURL::getSizeEstimateBeforeLimit() {
+  // This Operation will always be the only operation in a query tree, so we
+  // don't really need estimates.
+  return 100'000;
+}
+
+// _____________________________________________________________________________
+float LoadURL::getMultiplicity(size_t) {
+  // This Operation will always be the only operation in a query tree, so we
+  // don't really need estimates.
+  return 1;
+}
+
+// _____________________________________________________________________________
+bool LoadURL::knownEmptyResult() { return false; }
+
+// _____________________________________________________________________________
+std::unique_ptr<Operation> LoadURL::cloneImpl() const {
+  auto load = std::make_unique<LoadURL>(_executionContext, loadURLClause_);
+  load->cacheBreaker_ = cacheBreaker_;
+  return load;
+}
+
+// _____________________________________________________________________________
+vector<ColumnIndex> LoadURL::resultSortedOn() const { return {}; }
+
+// _____________________________________________________________________________
+Result LoadURL::computeResult(bool requestLaziness) {
+  try {
+    return computeResultImpl(requestLaziness);
+  } catch (const ad_utility::CancellationException&) {
+    throw;
+  } catch (const ad_utility::detail::AllocationExceedsLimitException&) {
+    throw;
+  } catch (const std::exception&) {
+    // If the `SILENT` keyword is set, catch the error and return a neutral
+    // Element.
+    if (loadURLClause_.silent_) {
+      IdTable idTable{getResultWidth(), getExecutionContext()->getAllocator()};
+      idTable.emplace_back();
+      for (size_t colIdx = 0; colIdx < getResultWidth(); ++colIdx) {
+        idTable(0, colIdx) = Id::makeUndefined();
+      }
```

Comment:
As you always have three elements you can do simply
`auto u = Id::makeUndefined(); idTable.push_back(std::array{u, u, u})`
which is simpler to read.

Reply:
Good though:

* 0 rows is the neutral element wrt UNION (like the "0" in real numbers)
* 1 row, but all columns are undefined e.g. because there are none is the neutral element wrt JOIN (like `1`).

This makes a large difference:
For Service we want the `1`. But for load we just don't want anything in the graph to be changed, so we have to return an empty result.....

By the way:
Do the UPDATe operations currently filter out triples that contain UNDEF? They definitely should, because otherwise all hell would break loose....
So please change this + comment + good catch.

Reply:
The Update already filters out `UNDEF`s. Rows with `UNDEF` are not considered for updates.